File: iOS/TUILiveKit/Sources/LiveStream/Main/Manager/Observer/LiveStreamObserver.swift
//
//  LiveStreamObserver.swift
//  TUILiveKit
//
//  Created by jeremiawang on 2024/11/21.
//

import Foundation
import LiveStreamCore
import RTCRoomEngine

class LiveStreamObserver: NSObject {
    private let manager: LiveStreamManager
    
    init(manager: LiveStreamManager) {
        self.manager = manager
        super.init()
    }
}

extension LiveStreamObserver: ConnectionObserver {
    func onConnectedUsersUpdated(userList: [TUIUserInfo], joinList: [TUIUserInfo], leaveList: [TUIUserInfo]) {
    }
    
    func onUserConnectionRequest(inviterUser: TUIUserInfo) {
    }
    
    func onUserConnectionCancelled(inviterUser: TUIUserInfo) {
    }
    
    func onUserConnectionAccepted(userInfo: TUIUserInfo) {
    }
    
    func onUserConnectionRejected(userInfo: TUIUserInfo) {
        manager.onUserConnectionRejected(userId: userInfo.userId)
    }
    
    func onUserConnectionTimeout(userInfo: TUIUserInfo) {
        manager.onUserConnectionTimeout(userId: userInfo.userId)
    }
    
    func onUserConnectionTerminated() {
        manager.onKickedOffSeat()
    }
    
    func onUserConnectionExited(userInfo: TUIUserInfo) {
        
    }
    
    func onConnectedRoomsUpdated(hostUserList: [TUIConnectionUser]) {
        manager.onConnectionUserListChanged(list: hostUserList)
    }
    
    func onCrossRoomConnectionRequest(hostUser: TUIConnectionUser) {
        
    }
    
    func onCrossRoomConnectionCancelled(hostUser: TUIConnectionUser) {
        
    }
    
    func onCrossRoomConnectionAccepted(hostUser: TUIConnectionUser) {
        
    }
    
    func onCrossRoomConnectionRejected(hostUser: TUIConnectionUser) {
        
    }
    
    func onCrossRoomConnectionTimeout(inviter: TUIConnectionUser, invitee: TUIConnectionUser) {
        
    }
    
    func onCrossRoomConnectionExited(hostUser: TUIConnectionUser) {
        
    }
    
    func onRoomDismissed(roomId: String) {
        
    }
}

File: Flutter/livekit/lib/live_stream/manager/observer/live_stream_observer.dart
import 'package:live_stream_core/live_core_widget/index.dart';
import 'package:tencent_live_uikit/live_stream/manager/module/co_host_manager.dart';

import '../live_stream_manager.dart';

class LiveStreamObserver extends ConnectionObserver {
  late final Context context;

  void init(Context context) {
    this.context = context;
  }

  LiveStreamObserver() {
    super.onConnectedUsersUpdated = (userList, joinList, leaveList) {};
    super.onUserConnectionRequest = (inviterUser) {};
    super.onUserConnectionCancelled = (inviterUser) {};
    super.onUserConnectionAccepted = (userInfo) {};
    super.onUserConnectionRejected = (userInfo) {
      context.coGuestManager.target?.onUserConnectionRejected(userInfo.userId);
    };
    super.onUserConnectionTimeout = (userInfo) {
      context.coGuestManager.target?.onUserConnectionTimeout(userInfo.userId);
    };
    super.onUserConnectionTerminated = () {
      context.coGuestManager.target?.onKickedOffSeat();
    };
    super.onUserConnectionExited = (userInfo) {};

    super.onConnectedRoomsUpdated = (hostUserList) {
      context.coHostManager.target?.onConnectionUserListChanged(hostUserList);
    };
    super.onCrossRoomConnectionRequest = (inviter) {
      context.coHostManager.target?.onConnectionRequestReceived(inviter);
    };
    super.onCrossRoomConnectionCancelled = (hostUser) {};
    super.onCrossRoomConnectionAccepted = (invitee) {
      context.coHostManager.target?.onConnectionRequestAccept(invitee);
    };
    super.onCrossRoomConnectionRejected = (invitee) {
      context.coHostManager.target?.onConnectionRequestReject(invitee);
    };
    super.onCrossRoomConnectionTimeout = (inviter, invitee) {
      context.coHostManager.target
          ?.onConnectionRequestTimeout(inviter, invitee);
    };
    super.onCrossRoomConnectionExited = (hostUser) {};
    super.onRoomDismissed = (roomId) {};
  }
}
